lib/connectors_sdk/office365/custom_client.rb

```diff
@@ -134,7 +134,10 @@ def list_items(drive_id, fields: [], break_after_page: false)
             yielded += 1
           end
 
-          if break_after_page && yielded >= 100 && stack.any?
+          if break_after_page && yielded >= 100
```

Comment:
@wangch079 I want to make sure I understood your concern about this logic, so let me know if I missed something here. As we paginate over items in a folder, we return that folder to the stack, relying on `item_children_next_link` to let us pick back up in the middle of that folder as needed.

Reply:
What I want to prevent here is the scenario when it's time to break the page and there's nothing to extract from the current drive.

E.g. that's a drive `foo` with one folder `bar`, within which there are more than 100 documents. When `item_children` is done, we should have `yielded >= 100`, and `stack` is empty. In such a case `[]` will be set to `page_cursor`, which, after I looked depper, will still work: In the next request, it will try to crawl drive `foo` again, but with empty `stack`, it won't enter the while loop and continue with the next drive.

I'm not holding strong opinions now, but if you want to remove `stack.any?`, remember to change this line

https://github.com/elastic/connectors/blob/8e942066309419a552b337e36c4580384f857bcc/lib/connectors_sdk/office365/extractor.rb#L54

to

```
if break_after_page && config.cursors.key?('page_cursor')
```

Reply:
Given the scenario of a drive `foo` with one folder `bar` that contains over 100 documents, after `item_children` is done iterating over the items in folder `bar`, `cursors['item_children_next_link']` will be _nil_. That will cause `list_items` to set `cursors['page_cursor']` to `[]`. Because `[].present?` evaluates to _false_, `yield_document_changes` will immediately proceed onto the next drive (or complete if `foo` was the only/last drive).

Am I missing anything else there?

Reply:
> Am I missing anything else there?

IMO, we should not store `[]` to `page_cursor`, and the presence of `page_cursor` indicates there's a page break (an empty stack should be considered present). Two negatives make a positive, so it works.

I'm also fine if we keep it this way.

BTW, you should check `item_children_next_link` here to determine whether we should break the page

https://github.com/elastic/connectors/blob/8e942066309419a552b337e36c4580384f857bcc/lib/connectors_sdk/office365/extractor.rb#L54-L56

and here to determine whether this is the end of the extraction

https://github.com/elastic/connectors/blob/8e942066309419a552b337e36c4580384f857bcc/lib/connectors_sdk/office365/extractor.rb#L59-L63

Reply:
@wangch079 Can I get one last review? I've changed the logic to avoid setting `:page_cursor => []`, and I've added a check for `item_children_next_link`. Based on my understanding of the cursor logic, I _think_ the behavior is the same given how _page_cursor_ and _item_children_next_link_ relate to each other, but these changes certainly make the logical expectations more explicit.